### ISSUE TITLE
Hardcoded default port numbers prevent concurrent execution - consider testing if port is in use

### ISSUE THREAD

**Issue by garikAsplund** [fermyon/spin#2470]
I had a couple projects open at the same time and noticed that the default ports are hardcoded and thus collide.

<img width="403" alt="Screenshot 2024-04-24 at 9 18 38 PM" src="https://github.com/fermyon/spin/assets/111464359/dd470064-4bdb-4e93-950d-04d367cd7d91">
<img width="732" alt="Screenshot 2024-04-24 at 9 18 28 PM" src="https://github.com/fermyon/spin/assets/111464359/8b670ce3-0caa-44b7-a452-04b2314ce7bb">

I'm using spin 2.4.2.

I'd be willing to give this a shot tomorrow. The issue is in [crates/trigger-http/src/lib.rs](https://github.com/fermyon/spin/blob/main/crates/trigger-http/src/lib.rs). 

I propose amending the serve and serve_tls functions to go through ~10 ports if the one listening is in use. This would not just work for default_value but also if the "--listen ADDRESS" is also in use.

**Comment by lann** [comment]
I don't think we should do this by default; if an app starts successfully but the port is hosting a different app than expected you could end up with frustration at best and a security issue at worst.

From personal experience I also just think this would be very surprising behavior; do you have any examples of other tools that work this way?

**Comment by garikAsplund** [comment]
No worries. I'm used to webdev and [Vite has this feature](https://github.com/vitejs/vite/blob/f71ba5b94a6e862460a96c7bf5e16d8ae66f9fe7/packages/vite/src/node/http.ts#L141-L175). 

<img width="396" alt="Screenshot 2024-04-25 at 8 19 43 AM" src="https://github.com/fermyon/spin/assets/111464359/1d3ca385-f0a8-458e-8c4c-e2c516ab33d0">


It might not be that common of an issue and it's an easy enough quickfix by adding the --listen flag.

**Comment by lann** [comment]
> Vite has this feature.

I've used Vite and never noticed this! :sweat_smile:

As pointed out out-of-band, another option that sort-of works today is `spin up --listen localhost:0`, which picks a random unused port. I say "sort-of" because it doesn't currently log the actual selected port, it just says `http://127.0.0.1:0` which isn't very helpful. _That_ is definitely something we should try to fix.

**Comment by lann** [comment]
https://github.com/fermyon/spin/issues/2471

**Comment by garikAsplund** [comment]
Thanks! I'll take a look.

**Comment by itowlson** [comment]
Given #2471 and #2473, is it okay to close this as addressed via the port 0 workaround?

**Comment by garikAsplund** [comment]
I believe they're separate issues--it's a matter of preference for changing the logic to increment if a port, likely 3000, is in use. @lann @itowlson I'll leave that to you to decide.

**Comment by lann** [comment]
I'm somewhat opposed to this behavior being default - `spin up` is used in production in some places (for better or worse...) - but I wouldn't be opposed to an opt-in feature (flag, env var and/or runtime config option).

**Comment by itowlson** [comment]
Fixed via #3182
